orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
       .def("reserve", [](std::vector<OrtValue>* v, const size_t len) { v->reserve(len); })
       .def("shrink_to_fit", [](std::vector<OrtValue>* v) { v->shrink_to_fit(); })
       .def("__len__", [](const std::vector<OrtValue>& v) { return v.size(); })
-      .def("__iter__", [](const std::vector<OrtValue>& v) {
-        return py::make_iterator(v.cbegin(), v.cend());
-      },
-           py::keep_alive<0, 1>())
+      .def(
+          "__iter__", [](const std::vector<OrtValue>& v) {
+            return py::make_iterator(v.cbegin(), v.cend());
+          },
+          py::keep_alive<0, 1>())
       .def("__getitem__", [](const std::vector<OrtValue>& v, const size_t idx) {
         return v.at(idx);
       })
       .def("dlpack_at", [](std::vector<OrtValue>* v, const size_t idx) {
         return py::reinterpret_steal<py::object>(ToDlpack(v->at(idx)));
-      });
+      })
+      .def("to_dlpack", [](const std::vector<OrtValue>& v, py::object to_tensor) -> py::list {
```

Comment:
Actually I don't think the sharing one OrtDLManagedTensor works. The main purpose of OrtDLManagedTensor is to hold the "OrtValue handle"'s reference. It's passed to torch tensor by DLManagedTensor's manager_ctx and will be destroyed when torch tensor release by calling the DLManagedTensor's deleter. The destroy will decrease the ref count of that OrtValue. By saying passing tensor owership from ORT to PyTorch, we actually means passing the reference of the OrtValue from ORT to PyTorch (not the DLManagedTensor). Using only one OrtDLManagedTensor cannot achieve this as one OrtDLManagedTensor can hold only one Ortvalue. The idea here has two problem:

1) when using only one OrtDLManagedTensor, the OrtDLManagedTensor's handle (which is OrtValue type) is replaced by new value for each asignment, means that the ref count for previous OrtValue is decreased by 1. And the final one's ref count will also be decreased by 1 after this function when that single OrtDLManagedTensor is destroyed. If in the main thread we unref all these OrtValues (suppose it's the only place to hold them), then all the OrtValue will be destroyed, then the returned torch tensor will be invalid because they point to invalid data address. I.e., this idea actually fails to pass the ownership of OrtValue from ORT to PyTorch.

2) All the DLManagedTensor passed to different torch tensors have the same deleter, which will try to delete the same single OrtDLManagedTensor. But actually this shared OrtDLManagedTensor is destroyed when we go out of this function. So when we finish using those torch tensors and try to destroy them, the deleter will be called, all those deleter will try to delete the same pointer, an invalid pointer pointed to a destroyed OrtDLManagedTensor.

I don't think a single OrtDLManagedTensor can solve this problem. Another idea is an OrtDLManagedTensor pool. During the first step, we still need to create one OrtDLManagedTensor for each OrtValue output, but these OrtDLManagedTensor can be shared start from step 2 (if we think this allocation is a real problem).

Reply:
You are right, I missed the fact the new tensor was responsible to delete the dlpack structure. So my local dlapck structure made onnxruntime crash because it could not be deleted. I modified the code to create a new dlpack structure for every tensor consuming it but I think it is still possible to create only one capsule instead of a capsule per dlpack structure.